Potentially bad assertion in ClasspathUtil

During IntelliJ import tests, the ClasspathUtil is failing with an AssertionError:

<details>
<summary>Stacktrace</summary>

```text
Caused by: java.lang.AssertionError
	at org.gradle.internal.classloader.ClasspathUtil.getClasspathForResource(ClasspathUtil.java:136)
	at org.gradle.internal.classloader.TransformingClassLoader.findClass(TransformingClassLoader.java:64)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at kotlin.reflect.jvm.internal.impl.descriptors.runtime.components.ReflectKotlinClassFinder.<init>(ReflectKotlinClassFinder.kt:30)
	at kotlin.reflect.jvm.internal.impl.descriptors.runtime.components.RuntimeModuleData$Companion.create(RuntimeModuleData.kt:65)
	at kotlin.reflect.jvm.internal.ModuleByClassLoaderKt.getOrCreateModule(moduleByClassLoader.kt:58)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:36)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:33)
	at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
	at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:31)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data.getModuleData(KDeclarationContainerImpl.kt)
	at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:50)
	at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:47)
	at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
	at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:31)
	at kotlin.reflect.jvm.internal.KClassImpl$Data.getDescriptor(KClassImpl.kt)
	at kotlin.reflect.jvm.internal.KClassImpl.getDescriptor(KClassImpl.kt:182)
	at kotlin.reflect.jvm.internal.KClassImpl.getDescriptor(KClassImpl.kt:44)
	at kotlin.reflect.full.KClassifiers.createType(KClassifiers.kt:47)
	at kotlin.reflect.jvm.internal.ReflectionFactoryImpl.typeOf(ReflectionFactoryImpl.java:123)
	at kotlin.jvm.internal.Reflection.typeOf(Reflection.java:127)
	at org.jetbrains.kotlin.idea.gradleTooling.IdeaKotlinExtras.<clinit>(IdeaKotlinExtras.kt:52)
	at org.jetbrains.kotlin.idea.gradleTooling.IdeaKotlinExtras$Surrogate.readResolve(IdeaKotlinExtras.kt:42)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1190)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at java.base/java.util.HashMap.readObject(HashMap.java:1552)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:647)
	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$InvocationHandlerImpl.readObject(ProtocolToModelAdapter.java:381)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at java.base/java.util.HashMap.readObject(HashMap.java:1552)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at org.gradle.tooling.internal.provider.serialization.PayloadSerializer.deserialize(PayloadSerializer.java:76)
```
</details>

Attached debugger at failing position:
<img width="1879" alt="image" src="https://user-images.githubusercontent.com/34319766/214094595-e9b0632c-96b8-4642-89f7-71adf08006fb.png">

The conditoin
```java
schemeSpecificPart.substring(pos + 1).endsWith("/" + name)
```

would be true, however since the string 'name' is prefixed by '/META-INF/versions/9' the assertion fails.

### Expected Behavior
As disabling the assertion leads to the no user visible issues, I would expect no assertion to be thrown.


### Context
This affects Gradle Sync tests inside IntelliJ

--- a/subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java
+++ b/subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java
@@ -133,7 +133,7 @@ public class ClasspathUtil {
                 String schemeSpecificPart = location.getRawSchemeSpecificPart();
                 int pos = schemeSpecificPart.indexOf("!");
                 if (pos > 0) {
-                    assert schemeSpecificPart.substring(pos + 1).equals("/" + name);
+                    assert schemeSpecificPart.substring(pos + 1).endsWith(name);
                     URI jarFile = new URI(schemeSpecificPart.substring(0, pos));
                     if (jarFile.getScheme().equals("file")) {
                         return new File(jarFile.getPath());

--- a/subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java
+++ b/subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java
@@ -133,7 +133,7 @@ public class ClasspathUtil {
                 String schemeSpecificPart = location.getRawSchemeSpecificPart();
                 int pos = schemeSpecificPart.indexOf("!");
                 if (pos > 0) {
-                    assert schemeSpecificPart.substring(pos + 1).endsWith(name);
+                    assert schemeSpecificPart.substring(pos + 1).endsWith("/" + name);
                     URI jarFile = new URI(schemeSpecificPart.substring(0, pos));
                     if (jarFile.getScheme().equals("file")) {
                         return new File(jarFile.getPath());

--- a/subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy
+++ b/subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy
@@ -22,7 +22,8 @@ import spock.lang.Specification
 class ClasspathUtilTest extends Specification {
     private static final URL JAR_URL = new URL('jar:file:home/duke/duke.jar!/')
     private static final URL HTTP_URL = new URL('http://www.foo.com/bar/')
-    private static final URL FILE_URL = new File('home/duke/duke.jar').toURI().toURL()
+    private static final File FILE = new File('/home/duke/duke.jar')
+    private static final URL FILE_URL = FILE.toURI().toURL()
     def factory = new DefaultClassLoaderFactory()
 
     def "filters non-file URLs from classpath"() {
@@ -46,7 +47,7 @@ class ClasspathUtilTest extends Specification {
         def file = ClasspathUtil.getClasspathForResource(jarUrlWithResourceName, "Test.class")
 
         then:
-        "/home/duke/duke.jar" == file.getPath()
+        FILE.getPath() == file.getPath()
     }
 
     @Issue("https://github.com/gradle/gradle/issues/23625")
@@ -58,6 +59,6 @@ class ClasspathUtilTest extends Specification {
         def file = ClasspathUtil.getClasspathForResource(jarUrlWithResourceName, "Test.class")
 
         then:
-        "/home/duke/duke.jar" == file.getPath()
+        FILE.getPath() == file.getPath()
     }
 }

--- a/subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy
+++ b/subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy
@@ -16,6 +16,7 @@
 package org.gradle.internal.classloader
 
 import org.gradle.internal.classpath.DefaultClassPath
+import spock.lang.Issue
 import spock.lang.Specification
 
 class ClasspathUtilTest extends Specification {
@@ -35,5 +36,28 @@ class ClasspathUtilTest extends Specification {
         then:
         !classpath.asURLs.any { it == JAR_URL || it == HTTP_URL }
         classpath.asURLs.any { it == FILE_URL }
+    }
+
+    def "getClasspathForResource for jar scheme with resource name"() {
+        given:
+        def jarUrlWithResourceName = new URL('jar:file:/home/duke/duke.jar!/Test.class')
+
+        when:
+        def file = ClasspathUtil.getClasspathForResource(jarUrlWithResourceName, "Test.class")
+
+        then:
+        "/home/duke/duke.jar" == file.getPath()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/23625")
+    def "getClasspathForResource for jar scheme with prefix resource name"() {
+        given:
+        def jarUrlWithResourceName = new URL('jar:file:/home/duke/duke.jar!/META-INF/versions/9/Test.class')
+
+        when:
+        def file = ClasspathUtil.getClasspathForResource(jarUrlWithResourceName, "Test.class")
+
+        then:
+        "/home/duke/duke.jar" == file.getPath()
     }
 }

